### DISCUSSION TITLE
Exiftool in github actions

### DISCUSSION THREAD

**Discussion by woss** [exiftool/exiftool#271] [6. Developers]
Hi all, 

now you can use amazing ExifTool as github action. This is useful when building an application or library around exiftool and you want to ensure that certain things are done correctly, like fetching certain metadata, c2pa box tags, or anything else. 

you can include multiple versions of the ExifTool in different jobs. 

Here is the link ->>
https://github.com/marketplace/actions/exiftool-action

**Comment by boardhead** [comment]
1.  I don't know what actions do.
2. I don't have much time to figure it out.
3. I've only ever had one other person request this.
4. Your actions are 3 versions behind the current ExifTool version.

**Reply by woss** [reply]
Ok, I'll explain my reasoning.

1. in them you can run pretty much anything, mostly they are used to run automated tests, create and publish docs, etc...
2. that's fine, I am quite experienced in maintaining them, it's my daily job. 
3. I know
4. I intend to keep them in sync when the main (this) repo tests pass. For example https://github.com/exiftool/exiftool/issues/272 

Unfortunately, there is no way for me to automatically hook into your release process since you don't create them on github, so I check once a week if you release Exiftool or not. Then I sync my fork (https://github.com/woss/fork-exiftool/actions) that runs the tests and when they pass I release a new version in exiftool-actions


this all would be much simpler if this repo would use actions and a github release together with your manual one. 

Let me know if you have any more questions :)

**Comment by boardhead** [comment]
"this all would be much simpler if this repo would use actions and a github release together with your manual one."

I have a long-standing issue about doing a github release (https://github.com/exiftool/exiftool/issues/13), but I haven't had time to figure out how to do this remotely from the command line.  Do you know a command to do this?

When you say "if this repo would use actions", can you be more specific about how this is done?

- Phil

**Reply by woss** [reply]
> When you say "if this repo would use actions", can you be more specific about how this is done?

By actions in this context, I mean Github actions. In my fork, you can see that I have already configured rudimentary actions that always run. For example, in the case of the exiftool main repo, the scenario would be that if there is a PR the actions are run, tests, code listing, and anything you need to deliver quality code.  I can help you do that.

> I have a long-standing issue about doing a github release (https://github.com/exiftool/exiftool/issues/13), but I haven't had time to figure out how to do this remotely from the command line. Do you know a command to do this?

Since Exiftool doesn't need to be installed a zip file called `asset` is enough, this is also covered by my [`exiftool-action`](https://github.com/woss/exiftool-action). Essentially the flow is following:

1. PR is created, worked on and actions run, tests pass and you are happy with it.
2. you merge it to the master. Tests run again to make sure nothing is broken. 
3. with the github CLI you can do the following `gh release create` and then manually increase the number and submit the release
4. the tag is created ( this is the version, let's say 12.93 ). Now the release contains the assets that are zip files of the directory. This is how basic release works.


More advanced release can have the action that runs on the release, or that the action creates the release when you create the tag from main, in which case release assets contain the gzip file and the checksum.  

I hope I have explained it well enough. As I said, I can help you with this if you wish.

p.s. in order to help you it would be good to see your current release script :) 

D

**Comment by boardhead** [comment]
Thanks for the details, but I don't think I'm any closer to making headway here.

I run tests myself before each release.  The problem in 12.90 was unique and won't happen again -- it was due to a restructuring of my release scripts for the new package formats.

You say "gh release create", but I don't have gh installed and it fails to install with "brew install gh" on my MacOS 10.14 dev machine.

My release scripts are complex and depend heavily on the file structure and installed utilities on my dev machine, so it doesn't really make sense to share them.

- Phil

**Reply by woss** [reply]
What's missing? It would help me if you could explain to me how you see this flow happen.

Btw, exiftool-actions already has the releases for the exiftool for each version I published. Although it doesn't contain as much as your gzip file people can still download older exiftool release.

**Comment by mceachen** [comment]
Hey Phil! There are a ton of ways to cut releases using GitHub -- but I think a robust way to do it is to hook it into a workflow (a workflow is just a YAML file in a magic .github/actions directory).

When you're ready to release:

1. you click a button in the github ui (under the Actions tab, you click a workflow, and
2. the workflow runs all your tests.
3. If they all pass, it increments your version number, rebuilds whatever it needs to do, builds the tarball, and uploads it whereever you see fit.

Here's where that button would live (this is how I release my npm packages):

<img width="1093" height="932" alt="image" src="https://github.com/user-attachments/assets/c8fc7844-4f8f-4100-a95b-400f103a0bd5" />

Just to give you a taste, here's what one of my workflows looks like:

<img width="1368" height="1032" alt="image" src="https://github.com/user-attachments/assets/f6ee3745-4aee-4223-a7f4-a01b38e399d0" />


hth!

**Comment by boardhead** [comment]
Thanks, but my goal was to do this from the command line.

**Reply by mceachen** [reply]
GitHub has a `gh` tool that can do the equivalent of clicking the button.

I think the bigger question is what you'd get out of this -- GHA (GitHub Actions) is **rough** -- it was **not** designed sensibly (or really "designed" -- it's an amalgam of chewing gum and duct tape and bailing wire. If you're using an LLM, and make the robot do the setup, it's all behind the scenes.

If you're happy with your release script, i'd keep it.

The only thing that would be nice would be to have a "GitHub Release" for every version -- and that can be done with the `gh` CLI tool as well. https://cli.github.com/manual/gh_release_create -- you could even add the call to the end of your release script and have it automated (and skip GHA altogether)
